### PR TITLE
api: add new API to delete multiple objects.

### LINKED ISSUE #467: Implement RemoveObjects API to remove objects in batches
Refer minio-go for further API details.

### PR DESCRIPTION
**Pull request by balamurugana**
This patch adds new API

public Iterable<Result<DeleteError>> removeObject(final String bucketName,
    final Iterable<String> objectNames)

which uses "Delete Multiple Objects" ReST API
(http://docs.aws.amazon.com/AmazonS3/latest/API/multiobjectdeleteapi.html).

Fixes #467

### CODE REVIEW

**Comment by balamurugana** [comment]
The PR succeeds in `play.minio.io` and `s3.amazonaws.com`

File: api/src/main/java/io/minio/messages/ErrorResponse.java [hunk 1]
```diff
@@ -33,21 +33,21 @@
 @SuppressWarnings("unused")
 public class ErrorResponse extends XmlEntity {
   @Key("Code")
-  private String code;
+  protected String code;
```
Comment (harshavardhana) [inline api/src/main/java/io/minio/messages/ErrorResponse.java]: why is this changed in this PR? any specific reason?

File: api/src/main/java/io/minio/MinioClient.java [hunk 2]
```diff
@@ -1884,6 +1888,156 @@ public void removeObject(String bucketName, String objectName)
...
+
+    return errorList;
+  }
+
+
+  /**
+   * Removes multiple objects from a bucket.
+   *
+   * </p><b>Example:</b><br>
+   * <pre>{@code minioClient.removeObject("my-bucketname", "my-objectname"); }</pre>
```
Comment (harshavardhana) [inline api/src/main/java/io/minio/MinioClient.java]: This might need an update to reflect to the new API changes.

File: functional/FunctionalTest.java [hunk 3]
```diff
@@ -435,6 +435,26 @@ public static void removeObject_test() throws Exception {
...
+    String[] filenames = new String[4];
+    for (int i = 0; i < 3; i++) {
+      String filename = createFile(1 * MB);
+      client.putObject(bucketName, filename, filename);
+      Files.delete(Paths.get(filename));
+      filenames[i] = filename;
+    }
+    filenames[3] = "nonexistent-object";
+
+    for (Result<?> r : client.removeObject(bucketName, Arrays.asList(filenames))) {
```
Comment (harshavardhana) [inline functional/FunctionalTest.java]: Is this the right style of usage of this newer API? i thought we are supposed to send an iterator down not an Array list.

File: api/src/main/java/io/minio/messages/ErrorResponse.java (hunk 1 above)
Comment (balamurugana) [inline api/src/main/java/io/minio/messages/ErrorResponse.java]: Yes. Check `DeleteError.java`

File: api/src/main/java/io/minio/MinioClient.java (hunk 2 above)
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java]: Done

File: functional/FunctionalTest.java (hunk 3 above)
Comment (balamurugana) [inline functional/FunctionalTest.java]: The API requires `Iterable<String>`.  `Arrays,asList()` creates an `iterable` object.

File: api/src/main/java/io/minio/MinioClient.java [hunk 4]
```diff
@@ -1884,6 +1888,157 @@ public void removeObject(String bucketName, String objectName)
...
+           InvalidKeyException, NoResponseException, XmlPullParserException, ErrorResponseException,
+           InternalException {
+    Map<String,String> queryParamMap = new HashMap<>();
+    queryParamMap.put("delete", "");
+
+    DeleteRequest request = new DeleteRequest(objectList);
+    HttpResponse response = executePost(bucketName, null, null, queryParamMap, request);
+
+    String bodyContent = "";
+    Scanner scanner = new Scanner(response.body().charStream());
```
Comment (vadmeste) [inline api/src/main/java/io/minio/MinioClient.java]: I am not sure what's the purpose of this scanner? can you add a comment ?

File: api/src/main/java/io/minio/MinioClient.java [hunk 5]
```diff
@@ -1884,6 +1888,157 @@ public void removeObject(String bucketName, String objectName)
   }
 
 
+  private List<DeleteError> removeObject(String bucketName, List<DeleteObject> objectList)
```
Comment (vadmeste) [inline api/src/main/java/io/minio/MinioClient.java]: Do you mean removeObjects() ? since it takes a list of delete object requests.

File: api/src/main/java/io/minio/MinioClient.java (hunk 4 above)
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java]: Done

File: api/src/main/java/io/minio/MinioClient.java (hunk 5 above)
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java]: Nope.  I wanted to keep a overloaded `removeObject()`

File: api/src/main/java/io/minio/MinioClient.java (hunk 4 above)
Comment (vadmeste) [inline api/src/main/java/io/minio/MinioClient.java]: > Its already commented below.

yeah I didn't expect all these lines are just to read the body, there is not an easier way? ``// Use scanner to read the entire body`` will be a better comment probably.

File: api/src/main/java/io/minio/MinioClient.java [hunk 6]
```diff
@@ -1884,6 +1888,156 @@ public void removeObject(String bucketName, String objectName)
...
+      }
+    } finally {
+      response.body().close();
+      scanner.close();
+    }
+
+    List<DeleteError> errorList = null;
+
+    bodyContent = bodyContent.trim();
+    DeleteError error = new DeleteError(new StringReader(bodyContent));
```
Comment (vadmeste) [inline api/src/main/java/io/minio/MinioClient.java]: DeleteError as the S3 only returns errors ?

File: api/src/main/java/io/minio/MinioClient.java (hunk 4 above)
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java]: Done.

Many ways to do that http://stackoverflow.com/questions/309424/read-convert-an-inputstream-to-a-string

File: api/src/main/java/io/minio/MinioClient.java (hunk 6 above)
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java]: `DeleteError` is multi object removal error in various cases as per the spec.

Comment (vadmeste) [inline api/src/main/java/io/minio/MinioClient.java]: > DeleteError is multi object removal error in various cases as per the spec.

yes thanks, the spec is little confusing, especially how it is returning 200 OK also for errors

Comment (vadmeste) [inline api/src/main/java/io/minio/MinioClient.java]: Can you also add a comment for this, to say that we need sometimes to parse the body twice since the xml tree structure is not the same in some circumstances.

Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java]: Done
